src/NumericalAlgorithms/Integration/GslQuadAdaptive.hpp

```diff
@@ -0,0 +1,263 @@
+// Distributed under the MIT License.
+// See LICENSE.txt for details.
+
+#pragma once
+
+#include <gsl/gsl_integration.h>
+#include <memory>
+
+#include "Utilities/Gsl.hpp"
+
+/// \cond
+namespace PUP {
+class er;
+}  // namespace PUP
+/// \endcond
+
+/// Numerical integration algorithms
+namespace integration {
+
+namespace detail {
+
+// The GSL functions require the integrand to have this particular function
+// signature. In particular, any extra parameters to the functions must be
+// passed as a void* and re-interpreted appropriately.
+template <typename IntegrandType>
+double integrand(double x, void* params) {
+  const auto function = reinterpret_cast<IntegrandType*>(params);
+  return (*function)(x);
+}
+
+class GslQuadAdaptiveImpl {
+ public:
+  explicit GslQuadAdaptiveImpl(size_t max_intervals) noexcept;
+
+  GslQuadAdaptiveImpl() = default;
+  GslQuadAdaptiveImpl(const GslQuadAdaptiveImpl&) = delete;
+  GslQuadAdaptiveImpl& operator=(const GslQuadAdaptiveImpl&) = delete;
+  GslQuadAdaptiveImpl(GslQuadAdaptiveImpl&&) noexcept = default;
+  GslQuadAdaptiveImpl& operator=(GslQuadAdaptiveImpl&& rhs) noexcept = default;
+  ~GslQuadAdaptiveImpl() noexcept = default;
+
+  // clang-tidy: no runtime references
+  void pup(PUP::er& p) noexcept;  // NOLINT(google-runtime-references)
+
+ private:
```

Comment:
fix order of public, protected, private https://google.github.io/styleguide/cppguide.html#Declaration_Order

Reply:
If I change the order, in `workspace_` I get that `gsl_integration_workspace_deleter` hasn't been declared yet. I don't know how to comply with the Declaration_Order then, do you have an idea? Thanks in advance!

Reply:
I now moved everything, but the `gsl_integration_workspace_deleter`.

Reply:
Sorry this one took me so long to get back on. I wanted to think about it and play around with your code a bit to understand the situation. As I see it there are a few options:
1. Leave as-is
2. Have `const auto& get_workspace() ...` function that is protected and returns the private member
3. Have `gsl_integration_workspace_deleter` (which should actually be CamelCase since it's a class) be in the `protected` section
I have no strong feelings for or against any of them. 1. and 3. definitely result in the least cumbersome code (and my view is if adhering to ordering conventions makes the code harder to maintain then it's better to break convention). Since the deleter is technically used by the derived classes I prefer 3, but would like @wthrowe's opinion on this before you make a change :)

Reply:
I also like 3 more than 1.

A possible variation on 2 would be a protected `gsl_integration_workspace* get_workspace()`, so the derived classes never have to deal with the unique_ptr at all.

Reply:
Thank you for your suggestions. I for now went with option 3, if you prefer option 2 though please let me know.